src/Lean/Elab/Term.lean

```diff
@@ -1675,6 +1681,9 @@ If `errToSorry` is true, then if coercion insertion fails, this function returns
 Otherwise, it throws the error.
 -/
 def elabTermEnsuringType (stx : Syntax) (expectedType? : Option Expr) (catchExPostpone := true) (implicitLambda := true) (errorMsgHeader? : Option String := none) : TermElabM Expr := do
+  if let some expectedType := expectedType? then
+    if (← pure (debug.sorryProps.get (← getOptions))) && !expectedType.hasMVar && (← isProp expectedType) then
+      return ← mkSorry expectedType false
```

Comment:
For `hasMVar`, there could be an `instantiateMVars` to avoid some false negatives.
```suggestion
  if debug.sorryProps.get (← getOptions) then
    if let some expectedType := expectedType? then
      let expectedType ← instantiateMVars expectedType
      if ← pure !expectedType.hasMVar <&&> isProp expectedType then
        return ← mkSorry expectedType false
```